src/main/java/com/limechain/babe/BlockProductionVerifier.java

```diff
@@ -39,19 +55,45 @@ public boolean verifyAuthorship(BlockHeader blockHeader, EpochData currentEpochD
         int authorityIndex = babePreDigest.map(BabePreDigest::getAuthorityIndex).map(Math::toIntExact).get();
         Authority verifyingAuthority = authorities.get(authorityIndex);
 
-        VrfOutputAndProof vrfOutputAndProof = VrfOutputAndProof.wrap(
-                babePreDigest.get().getVrfOutput(), babePreDigest.get().getVrfProof());
-        VerifySignature signature = new VerifySignature(
-                signatureData, blockHeader.getHashBytes(), verifyingAuthority.getPublicKey(), Key.SR25519);
-
-        return Sr25519Utils.verifySignature(signature) &&
-                verifySlotWinner(authorityIndex,
-                        authorities,
-                        currentEpochIndex,
-                        randomness,
-                        descriptor,
-                        currentSlotNumber,
-                        vrfOutputAndProof);
+        if (!checkBlockEquivocation(verifyingAuthority, blockHeader, runtime, currentSlotNumber)) {
+            return false;
+        }
+
+        VrfOutputAndProof vrfOutputAndProof = VrfOutputAndProof.wrap(babePreDigest.get().getVrfOutput(), babePreDigest.get().getVrfProof());
+        VerifySignature signature = new VerifySignature(signatureData, blockHeader.getHashBytes(), verifyingAuthority.getPublicKey(), Key.SR25519);
+
+        boolean isVerified = Sr25519Utils.verifySignature(signature) &&
+                verifySlotWinner(authorityIndex, authorities, currentEpochIndex, randomness, descriptor, currentSlotNumber, vrfOutputAndProof);
+
+        if (isVerified) {
+            currentSlotAuthorBlockMap.put(StringUtils.toHexWithPrefix(verifyingAuthority.getPublicKey()),blockHeader);
+        }
+
+        return isVerified;
+    }
+
+    private boolean checkBlockEquivocation(Authority authority,
+                                           BlockHeader blockHeader,
+                                           Runtime runtime,
+                                           BigInteger currentSlotNumber) {
+        String hexPublicKey = HexUtils.toHexString(authority.getPublicKey());
+        if (currentSlotAuthorBlockMap.containsKey(hexPublicKey)) {
+            BlockHeader firstBlockHeader = currentSlotAuthorBlockMap.get(hexPublicKey);
+            if (!firstBlockHeader.equals(blockHeader)) {
+                BlockEquivocationProof blockEquivocationProof = new BlockEquivocationProof();
+                blockEquivocationProof.setPublicKey(authority.getPublicKey());
+                blockEquivocationProof.setSlotNumber(currentSlotNumber);
+                blockEquivocationProof.setFirstBlockHeader(firstBlockHeader);
+                blockEquivocationProof.setSecondBlockHeader(blockHeader);
+
+                Optional<OpaqueKeyOwnershipProof> opaqueKeyOwnershipProof = runtime.generateKeyOwnershipProof(currentSlotNumber, authority.getPublicKey());
```

Comment:
We should try and keep our code lines shorter than the suggested length limit (vertical line in intellij). This will help improve code readability on different monitors and get us closer to a uniform code styling.